### PR TITLE
API-1835: Migrate multiple controllers to SSA

### PR DESCRIPTION
**Pull request by bertinatto**
Better reviewed commit by commit.

Proof: https://github.com/openshift/cluster-authentication-operator/pull/721

Verb usage comparison between single-node jobs from proof and https://github.com/openshift/cluster-authentication-operator/pull/715:

verb_usage_comparison_auth

### CODE REVIEW

File: pkg/operator/managementstatecontroller/management_state_controller.go [hunk 1]
```diff
@@ -82,11 +85,5 @@ func (c ManagementStateController) sync(ctx context.Context, syncContext factory
 	}
 
 	status := applyoperatorv1.OperatorStatus().WithConditions(cond)
-	if applyError := c.operatorClient.ApplyOperatorStatus(ctx, c.controllerInstanceName, status); applyError != nil {
-		if err == nil {
-			return applyError
-		}
```
Comment (bertinatto) [inline pkg/operator/managementstatecontroller/management_state_controller.go:83]: This seemed like an oversight.

File: pkg/operator/staticpod/controllers.go [hunk 2]
```diff
@@ -359,7 +359,7 @@ func (b *staticPodOperatorControllerBuilder) ToControllers() (manager.Controller
 		eventRecorder,
 	).AddKubeInformers(b.kubeInformers), 1)
 
-	manager.WithController(unsupportedconfigoverridescontroller.NewUnsupportedConfigOverridesController(b.staticPodOperatorClient, eventRecorder), 1)
+	manager.WithController(unsupportedconfigoverridescontroller.NewUnsupportedConfigOverridesController("UnsupportedConfigOverridesController", b.staticPodOperatorClient, eventRecorder), 1)
```
Comment (deads2k) [inline pkg/operator/staticpod/controllers.go:362]: shouldn't instanceName be more like `b.operatorName`

File: pkg/operator/apiserver/controllerset/apiservercontrollerset.go [hunk 3]
```diff
@@ -105,7 +105,7 @@ func NewAPIServerControllerSet(
 // WithConfigUpgradableController adds a controller for the operator to check for presence of
 // unsupported configuration and to set the Upgradable condition to false if it finds any
 func (cs *APIServerControllerSet) WithConfigUpgradableController() *APIServerControllerSet {
-	cs.configUpgradableController.controller = unsupportedconfigoverridescontroller.NewUnsupportedConfigOverridesController(cs.operatorClient, cs.eventRecorder)
+	cs.configUpgradableController.controller = unsupportedconfigoverridescontroller.NewUnsupportedConfigOverridesController("UnsupportedConfigOverridesController", cs.operatorClient, cs.eventRecorder)
```
Comment (deads2k) [inline pkg/operator/apiserver/controllerset/apiservercontrollerset.go:108]: shouldn't instanceName be more like cs.name

File: pkg/operator/encryption/controllers.go [hunk 4]
```diff
@@ -65,6 +65,7 @@ func NewControllers(
 	return &Controllers{
 		controllers: []runner{
 			controllers.NewKeyController(
+				"EncryptionKeyController",
```
Comment (deads2k) [inline pkg/operator/encryption/controllers.go:68]: same questions about the instance names here.

File: pkg/operator/encryption/controllers/migration_controller.go [hunk 5]
```diff
@@ -103,46 +104,65 @@ func NewMigrationController(
...
+	// Status for these conditions is left out to make sure it's correctly set in every branch
+	degradedCondition := applyoperatorv1.OperatorCondition().
+		WithType("EncryptionMigrationControllerDegraded")
+	progressingCondition := applyoperatorv1.OperatorCondition().
+		WithType("EncryptionMigrationControllerProgressing")
+
 	defer func() {
 		if degradedCondition == nil && progressingCondition == nil {
 			return
 		}
```
Comment (deads2k) [inline pkg/operator/encryption/controllers/migration_controller.go:122]: if one is nil and not the other, fail.

File: pkg/operator/encryption/controllers.go (hunk 4 above)
Comment (bertinatto) [reply pkg/operator/encryption/controllers.go]: Fixed all of them

File: pkg/operator/encryption/controllers/migration_controller.go (hunk 5 above)
Comment (bertinatto) [reply pkg/operator/encryption/controllers/migration_controller.go]: Makes sense, if one of these is nil, _WithConditions_ will panic, so I made this an OR instead

File: pkg/operator/encryption/controllers/key_controller.go [hunk 6]
```diff
@@ -252,7 +265,7 @@ func (c *keyController) generateKeySecret(keyID uint64, currentMode state.Mode,
 		InternalReason: internalReason,
 		ExternalReason: externalReason,
 	}
-	return secrets.FromKeyState(c.component, ks)
+	return secrets.FromKeyState(c.instanceName, ks)
```
Comment (deads2k) [inline pkg/operator/encryption/controllers/key_controller.go:268]: This change adjusts the name of secrets we create, which (I think) will impact upgrade and downgrade with this feature one.  Please keep this as `c.component`.

File: pkg/operator/encryption/controllers/migration_controller.go [hunk 7]
```diff
@@ -230,7 +249,7 @@ func (c *migrationController) migrateKeysIfNeededAndRevisionStable(ctx context.C
...
 
 		// update secret annotations
-		oldWriteKey, err := secrets.FromKeyState(c.component, grActualKeys.WriteKey)
+		oldWriteKey, err := secrets.FromKeyState(c.instanceName, grActualKeys.WriteKey)
```
Comment (deads2k) [inline pkg/operator/encryption/controllers/migration_controller.go:252]: this one too.  We'll not identify old on upgrade, right?

File: pkg/operator/encryption/controllers/state_controller.go [hunk 8]
```diff
@@ -152,7 +162,7 @@ func (c *stateController) generateAndApplyCurrentEncryptionConfigSecret(ctx cont
 }
 
 func (c *stateController) applyEncryptionConfigSecret(ctx context.Context, encryptionConfig *apiserverconfigv1.EncryptionConfiguration, recorder events.Recorder) (bool, error) {
-	s, err := encryptionconfig.ToSecret("openshift-config-managed", fmt.Sprintf("%s-%s", encryptionconfig.EncryptionConfSecretName, c.component), encryptionConfig)
+	s, err := encryptionconfig.ToSecret("openshift-config-managed", fmt.Sprintf("%s-%s", encryptionconfig.EncryptionConfSecretName, c.instanceName), encryptionConfig)
```
Comment (deads2k) [inline pkg/operator/encryption/controllers/state_controller.go:165]: and this one.

File: pkg/operator/encryption/controllers/key_controller.go (hunk 6 above)
Comment (deads2k) [reply pkg/operator/encryption/controllers/key_controller.go:268]: > This change adjusts the name of secrets we create, which (I think) will impact upgrade and downgrade with this feature one. Please keep this as `c.component`.

Oh, I see. The name of the struct var changed, but the value did not.

**Pull request merged**